File: ShopifySharp.GraphQL.Parser/Parser.cs
using System.IO.Pipelines;
using System.Runtime.CompilerServices;
using System.Text;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;

namespace ShopifySharp.GraphQL.Parser;

public class Parser(CasingType propertyNameCasingType)
{
    public async IAsyncEnumerable<GeneratedFile> ParseAsync(ReadOnlyMemory<char> graphqlData, [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        using var stream = new MemoryStream();
        await using var streamWriter = new StreamWriter(stream, Encoding.UTF8);
        var pipe = new Pipe();

        var visitor = new ShopifyGraphAstVisitor();
        var context = new WriterContext(pipe.Writer, cancellationToken)
        {
            CasingType = propertyNameCasingType
        };

        // Start reading the pipe before writing to it, or else the writer can hang if the pipe fills before it's read
        var readTask = ReadPipeAsync(pipe.Reader, cancellationToken);

        var ast = GraphQLParser.Parser.Parse(graphqlData);
        var writeTask = visitor.VisitDocumentAsync(ast, context);
        await writeTask.ConfigureAwait(false);
        await pipe.Writer.CompleteAsync().ConfigureAwait(false);

        var csharpCode = await readTask.ConfigureAwait(false);

        // Parse the generated code into individual files
        await foreach (var item in ParseCsharpToGeneratedFilesAsync(csharpCode, cancellationToken))
            yield return item;
    }

    private static async IAsyncEnumerable<GeneratedFile> ParseCsharpToGeneratedFilesAsync(string csharpCode, [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        var csharpTree = await CSharpSyntaxTree.ParseText(csharpCode, cancellationToken: cancellationToken)
            .GetRootAsync(cancellationToken);
        var syntaxRoot = (CompilationUnitSyntax)csharpTree;
        var usings = syntaxRoot.Usings;
        var externals = syntaxRoot.Externs;

        foreach (var ns in syntaxRoot.Members.OfType<BaseNamespaceDeclarationSyntax>())
        {
            foreach (var type in ns.Members.OfType<BaseTypeDeclarationSyntax>())
            {
                var unit = CompilationUnit()
                    .WithExterns(externals)
                    .WithUsings(usings)
                    .AddMembers(
                        ns.WithMembers(
                            SingletonList<MemberDeclarationSyntax>(type))
                    )
                    .NormalizeWhitespace(eol: Environment.NewLine);

                yield return new GeneratedFile(type.Identifier.Text + ".cs", unit.ToFullString());
            }
        }
    }

    private static async Task<string> ReadPipeAsync(PipeReader reader, CancellationToken ct = default)
    {
        var sb = new StringBuilder();

        while (true)
        {
            var result = await reader.ReadAsync(ct).ConfigureAwait(false);
            var buffer = result.Buffer;

            foreach (var segment in buffer)
                sb.Append(Encoding.UTF8.GetString(segment.Span));

            reader.AdvanceTo(buffer.End);

            if (result.IsCompleted || result.IsCanceled) break;
        }

        await reader.CompleteAsync().ConfigureAwait(false);
        return sb.ToString();
    }
}

File: ShopifySharp.GraphQL.Parser/Parser.fs
module ShopifySharp.GraphQL.Parser.Parser

open System
open System.IO.Pipelines
open System.Threading
open System.Threading.Tasks
open FSharp.Control
open GraphQLParser
open GraphQLParser.Visitors
open ShopifySharp.GraphQL.Parser.Utils

let private parseAsync (casing: Casing)
                       (assumeNullability: bool)
                       (graphqlData: ReadOnlyMemory<char>)
                       cancellationToken
                       : ValueTask<ParserContext> =
    let ast = Parser.Parse(graphqlData)
    let context = ParserContext(casing, assumeNullability, ast, cancellationToken)
    let visitor: ASTVisitor<ParserContext> = Visitor()

    // let sdlThing = StructurePrinter()
    // use writer = new StringWriter()
    // let result = sdlThing.PrintAsync(ast, writer, context.CancellationToken).AsTask()
    //              |> Async.AwaitTask
    //              |> Async.RunSynchronously

    ValueTask<ParserContext>(task {
        // TODO: It's simpler to just skip the visitor concept entirely and pass the parsed GraphDocument to the writers.
        //       The QueryBuilderWriter does this already; the VisitedTypeWriter should do it as well.
        do! visitor.VisitAsync(ast, context)
        return context
    })

let private parseAndWriteToFilesystem (parseAndWriteFn: ParserContext -> PipeWriter -> ValueTask)
                                      (destination: FileSystemDestination)
                                      (context: ParserContext)
                                      : ValueTask =
    let cancellationToken = context.CancellationToken

    ValueTask(task {
        let pipe = Pipe(PipeOptions())
        let readTask = (readPipe pipe.Reader cancellationToken).ConfigureAwait(false)

        do! parseAndWriteFn context pipe.Writer
        do! pipe.Writer.CompleteAsync().ConfigureAwait(false);

        let! csharpCode = readTask
        do! (FileSystem.writeCsharpCodeToFileSystem destination csharpCode cancellationToken).ConfigureAwait(false)
    })

let ParseAsync (casing: Casing)
               (assumeNullability: bool)
               (graphqlData: ReadOnlyMemory<char>)
               (cancellationToken: CancellationToken)
               : ValueTask<VisitedTypes[]> =
    ValueTask<VisitedTypes[]>(task {
        let! context = parseAsync casing assumeNullability graphqlData cancellationToken
        return context.GetVisitedTypes()
    })

let ParseAndWriteAsync (typesDestination: FileSystemDestination, servicesDestination: FileSystemDestination)
                       (casing: Casing)
                       (assumeNullability: bool)
                       (graphqlData: ReadOnlyMemory<char>)
                       cancellationToken
                       : ValueTask =
    ValueTask(task {
        let! context = parseAsync casing assumeNullability graphqlData cancellationToken
        do! parseAndWriteToFilesystem VisitedTypeWriter.writeVisitedTypesToPipe typesDestination context
        do! parseAndWriteToFilesystem QueryBuilderWriter.WriteQueryBuildersToPipe servicesDestination context
    })
